#66: [Persist emails before send] persist messages before send (closes #66)

calippo opened the pull request:
Closes #66

## Test Plan

### tests performed

> A Test Plan is used to show what you tested to make sure your code works fine. You should write here all that you did to test, and provide some results of your testing.

> These results can be screenshots, query results, or even just pointers to unit tests that successfully passed.

### tests not performed (domain coverage)

> At times not everything can be tested, and writing what hasn't been tested is just as important as writing what has been tested.

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.

calippo committed:
Ask for persistance

calippo committed:
Add persistance

calippo committed:
Move codecs away from persistance module

calippo commented on src/main/scala/mailo/persistence/EmailPersistanceActor.scala:
```diff
@@ -0,0 +1,50 @@
+    case command@CleanQueue(sequenceNr) =>
+      log.info("received event {}", command.toString)
+      saveSnapshot(state)
+      deleteMessages(sequenceNr)
```
honestly I'm not sure about the clean queue part, need to better test it

bytecodeguru commented on src/main/scala/mailo/persistence/CustomContentTypeCodecs.scala:
```diff
@@ -0,0 +1,20 @@
+    final def apply(c: HCursor): Decoder.Result[ContentType] =
+      for {
+        contentType <- c.downField("contentType").as[String]
+        result <- ContentType.parse(contentType).left.map(_ => DecodingFailure("Falied decoding content type", Nil))
```
Typo `Falied` -> `Failed`

bytecodeguru commented on src/main/scala/mailo/persistence/SnapshotHelper.scala:
```diff
@@ -0,0 +1,18 @@
+object SnapshotHelper extends LazyLogging {
+  val logSnapshotResult: PartialFunction[Any, Unit] = {
+    case SaveSnapshotSuccess(metadata) =>
+      logger.debug(s"Snaposhot successfully stored")
```
typo `Snaposhot` -> `Snapshot`

bytecodeguru reviewed (commented):
I need an offline explanation of the persistence machinery.

I've found just a couple of typos for now.

calippo committed:
Refactor persistance actor using event source approach

calippo committed:
Minor names and tests refactor

calippo committed:
Fix typo

calippo committed:
Add akka remote, remove akka query

calippo committed:
Add email sender module

calippo committed:
Refactor at most once mailo

calippo committed:
Minor at least once fixes

calippo committed:
Improve persistence actor

calippo committed:
Add reference conf with akka confs

calippo committed:
Add dry run, to be converted in integration tests

calippo commented:
@bytecodeguru as discussed offline, updated macro issue with the following:
- convert dry run to integration test
- add dead letter queue
- add queue monitoring api

https://github.com/buildo/mailo/issues/59

calippo committed:
Fork in test

calippo committed:
Fix test mail client

calippo committed:
Add dead letters handler and improve logic

calippo committed:
Handle dead email also on json decoding

calippo committed:
Let the user decide how to handle dead emails

calippo committed:
Improve errors handling

calippo committed:
Fix actor logging

calippo committed:
Update sbt-release-early

calippo committed:
Revert "Update sbt-release-early"

This reverts commit 22b4c51ff694f1ac8adc204b6871eaa643a32ba4.

bytecodeguru commented on src/main/resources/reference.conf:
```diff
@@ -0,0 +1,18 @@
+#       "mailo.persistence.SendEmail" = proto
+#     }
+#   }
+# }
```
Why is this commented out?

calippo replied:
It took me 1 hour to find the proper setting, and currently it's buggish xD

bytecodeguru commented on src/main/scala/mailo/AtLeastOnceMailo.scala:
```diff
@@ -1,40 +1,39 @@
-    }
+  def send(mail: Mail): Future[Either[MailError, MailResult]] = {
+    ask(emailPersistanceActor, SendEmail(mail))
+      .map(_ => Right(LocallyQueued))
```
Is it really safe to ignore `_`?

bytecodeguru commented on src/main/scala/mailo/DryRun.scala:
```diff
@@ -0,0 +1,55 @@
+    Future(Right(MailResponse("id", "ok")))
+}
+
+object AtLeastOnceDryRun extends App {
```
Is this a test?

calippo replied:
It's a dry run, not a proper test, this is removed in the next pr.

bytecodeguru commented on src/main/scala/mailo/persistence/CustomContentTypeCodecs.scala:
```diff
@@ -0,0 +1,20 @@
+import io.circe.DecodingFailure
+
+trait CustomContentTypeCodecs {
+  implicit val encodeFoo: Encoder[ContentType] = new Encoder[ContentType] {
```
What about `contentTypeEncoder` and `contentTypeDecoder`?

bytecodeguru commented on src/main/scala/mailo/persistence/EmailPersistanceActor.scala:
```diff
@@ -0,0 +1,82 @@
+    case EmailApplicativeErrorEvent(event, errorMessage) =>
+      log.error(s"%{event} failed with error: ${errorMessage}")
+    case EmailCommunicationErrorEvent(sendEmail, errorMessage) =>
+      log.error(s"${sendEmail} failed with error ${errorMessage}, rescheduling the message")
```
We are not rescheduling anything yet... the logged message is misleading.

bytecodeguru reviewed (changes requested):
Nice!
I've left some minor suggestions.

calippo committed:
Address minor style fixes

calippo commented:
@bytecodeguru addressed most of the comments.

this is the only major change
![image](https://user-images.githubusercontent.com/2476080/51403802-d3218100-1b51-11e9-962e-993b46204b64.png)

Q: "is it safe to discard the result"?
A: "we're now returning the result itself, there shouldn't be any problem (even though I'm not convinces `.mapTo` is super typesafe).

bytecodeguru reviewed (approved):
LGTM 👍
